code/_onclick/drag_drop.dm

```diff
@@ -47,4 +47,65 @@ actually if you do something in that proc like changing user location or whateve
 to inform the game this action was expected and its fine
 */
 /atom/proc/MouseDrop_T(atom/dropping, mob/user, params) // return TRUE if you want to prevent us click the object after it
-	return
+	return FALSE
+
+/client/MouseDown(datum/object, location, control, params)
+	if(QDELETED(object)) // Yep, you can click on qdeleted things before they have time to nullspace. Fun.
+		return
+	SEND_SIGNAL(src, COMSIG_CLIENT_MOUSEDOWN, object, location, control, params)
+	if(mouse_down_icon)
+		mouse_pointer_icon = mouse_down_icon
+	var/delay = mob.CanMobAutoclick(object, location, params)
+	if(delay)
+		selected_target[1] = object
+		selected_target[2] = params
+		while(selected_target[1])
+			Click(selected_target[1], location, control, selected_target[2])
+			sleep(delay)
+
+/client/MouseUp(object, location, control, params)
+	if(SEND_SIGNAL(src, COMSIG_CLIENT_MOUSEUP, object, location, control, params) & COMPONENT_CLIENT_MOUSEUP_INTERCEPT)
+		click_intercept_time = world.time
+	if(mouse_up_icon)
+		mouse_pointer_icon = mouse_up_icon
+	selected_target[1] = null
+
+/mob/proc/CanMobAutoclick(object, location, params)
+	return FALSE
+
+/mob/living/carbon/CanMobAutoclick(atom/object, location, params)
+	if(!object.IsAutoclickable())
+		return FALSE
+	var/obj/item/active_item = get_active_hand()
+	return active_item?.CanItemAutoclick(object, location, params)
+
+/obj/item/proc/CanItemAutoclick(object, location, params)
+	return FALSE
+
+/atom/proc/IsAutoclickable()
+	return TRUE
+
+/obj/screen/IsAutoclickable()
+	return FALSE
+
+/obj/screen/click_catcher/IsAutoclickable()
+	return TRUE
+
+/client/MouseDrag(src_object, atom/over_object, src_location, turf/over_location, src_control, over_control, params)
+	var/list/modifiers = params2list(params)
+	if(!drag_start) // If we're just starting to drag
+		drag_start = world.time
+		drag_details = modifiers.Copy()
+	mouseParams = params
+	mouse_location_UID = over_location?.UID()
+	mouse_object_UID = over_object?.UID()
```

Comment:
```suggestion
	if(istype(over_location))
		mouse_location_UID = over_location.UID()
	if(istype(over_object))
		mouse_object_UID = over_object.UID()
```
Byond doesn't always pass atoms/turfs here. Welcome to the wonderful world of screen objects!

here's the """"correct"""" arguments for this function
```
MouseDrag(src_object as null|atom in usr.client,\
          over_object as null|atom in usr.client,\
          src_location as null|turf|text in usr.client,\
          over_location as null|turf|text in usr.client,\
          src_control as text, over_control as text, params as text)
```
horrifying